internal/api/validation/individual.go

```
@@ -0,0 +1,92 @@
+package validation
+
+import (
+	"net/mail"
+	"time"
+
+	"github.com/nrc-no/notcore/internal/api"
+	"github.com/nrc-no/notcore/internal/containers"
+	"github.com/nrc-no/notcore/internal/validation"
+)
+
+func ValidateIndividual(i *api.Individual) validation.ErrorList {
+	return validateIndividual(i, nil)
+}
+
+func ValidateIndividualList(i *api.IndividualList) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+	itemsPath := validation.NewPath("items")
+	for i, individual := range i.Items {
+		allErrs = append(allErrs, validateIndividual(individual, itemsPath.Index(i))...)
+	}
+	return allErrs
+}
+
+func validateIndividual(i *api.Individual, p *validation.Path) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+	allErrs = append(allErrs, validateIndividualCountryID(i.CountryID, p.Child("countryId"))...)
+	allErrs = append(allErrs, validateBirthDate(i.BirthDate, p.Child("birthDate"))...)
+	allErrs = append(allErrs, validateIndividualDisplacementStatus(i.DisplacementStatus, p.Child("displacementStatus"))...)
+	allErrs = append(allErrs, validateIndividualGender(i.Gender, p.Child("gender"))...)
+	allErrs = append(allErrs, validateIndividualEmail(i.Email, p.Child("email"))...)
+	return allErrs
+}
+
+func validateBirthDate(birthDate *time.Time, path *validation.Path) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+	// birthDate is optional
+	if birthDate != nil {
+		if birthDate.After(time.Now()) {
+			allErrs = append(allErrs, validation.Invalid(path, birthDate, "birthdate cannot be in the future"))
+		}
+		if *birthDate == (time.Time{}) {
+			allErrs = append(allErrs, validation.Invalid(path, birthDate, "must be a valid date"))
+		}
+	}
+	return allErrs
+}
```

Currently when parsing the excel, if the birth date is in an invalid format it will error and fail. This means we fail at the first instance and won't reach this validation. 

My first thought was to set the field to `nil` when parsing an error and raise at the validation stage. But if `nil` is an allowed value then we're hiding errors in the upload.

Maybe this needs to be done cell-by-cell in the unmarshelling?

---

We must consider unmarshaling and validation as 2 different things. The unmarshaling will decode without validating. So we can have multiple formats (json,csv,excel) and a single validation logic. We can use a similar approach with appending to an `allErrs` list when we unmarshal, though in this case the errors will be those of type conversion errors rather than "business logic" validations

---

@neb42 I have a local spike on how to validate the `ListIndividualsOptions` aka `GetAllOptions` (should be renamed) that actually does this. Instead of failing early, it will append all encountered errors to a list and return that